src/map/pet.c

```diff
@@ -226,60 +274,66 @@ static int pet_sc_check(struct map_session_data *sd, int type)
 	return 0;
 }
 
+/**
+ * Updates a pet's hunger value and timer and updates the pet's intimacy value based on its new hunger value.
+ *
+ * @param tid The pet's current hunger timer value.
+ * @param tick The amount of ticks to add to the pet's hunger timer.
+ * @param id The pet's master's account ID.
+ * @param data Unused.
+ * @return 1 on failure, 0 on success.
+ *
+ */
 static int pet_hungry(int tid, int64 tick, int id, intptr_t data)
 {
+	int interval;
 	struct map_session_data *sd;
 	struct pet_data *pd;
-	int interval;
 
-	sd=map->id2sd(id);
-	if(!sd)
+	if ((sd = map->id2sd(id)) == NULL)
 		return 1;
 
-	if(!sd->status.pet_id || !sd->pd)
+	if (sd->status.pet_id == 0 || (pd = sd->pd) == NULL)
 		return 1;
 
-	pd = sd->pd;
-	if(pd->pet_hungry_timer != tid){
-		ShowError("pet_hungry_timer %d != %d\n",pd->pet_hungry_timer,tid);
-		return 0;
+	if (pd->pet_hungry_timer != tid) {
+		ShowError("pet_hungry_timer %d != %d\n", pd->pet_hungry_timer, tid);
+		return 1;
 	}
+
 	pd->pet_hungry_timer = INVALID_TIMER;
 
-	if (pd->pet.intimate <= 0)
-		return 1; //You lost the pet already, the rest is irrelevant.
+	if (pd->pet.intimate <= PET_INTIMACY_NONE)
+		return 1; /// You lost the pet already, the rest is irrelevant.
+
+	if (battle_config.pet_hungry_delay_rate != 100)
+		interval = pd->petDB->hungry_delay * battle_config.pet_hungry_delay_rate / 100;
+	else
+		interval = pd->petDB->hungry_delay;
 
 	pd->pet.hungry--;
-	/* Pet Autofeed */
+
+	/** Pet Autofeed */
 	if (battle_config.feature_enable_pet_autofeed != 0) {
-		if (pd->petDB->autofeed == 1 && pd->pet.autofeed == 1 && pd->pet.hungry <= 25) {
+		if (pd->petDB->autofeed == 1 && pd->pet.autofeed == 1 && pd->pet.hungry <= PET_HUNGER_HUNGRY)
 			pet->food(sd, pd);
-		}
 	}
 
-	if( pd->pet.hungry < 0 )
-	{
+	if (pd->pet.hungry <= PET_HUNGER_NONE) {
 		pet_stop_attack(pd);
-		pd->pet.hungry = 0;
+		pd->pet.hungry = PET_HUNGER_NONE;
 		pet->set_intimate(pd, pd->pet.intimate - battle_config.pet_hungry_friendly_decrease);
-		if( pd->pet.intimate <= 0 )
-		{
-			pd->pet.intimate = 0;
+
+		if (pd->pet.intimate <= PET_INTIMACY_NONE)
 			pd->status.speed = pd->db->status.speed;
-		}
+
 		status_calc_pet(pd, SCO_NONE);
-		clif->send_petdata(sd,pd,1,pd->pet.intimate);
+		clif->send_petdata(sd, pd, 1, pd->pet.intimate);
+		interval = 20000; /// While starving, it's every 20 seconds.
 	}
-	clif->send_petdata(sd,pd,2,pd->pet.hungry);
-
-	if(battle_config.pet_hungry_delay_rate != 100)
```

Comment:
with this code change the rate would not longer affect the special case of starving, please move it back to where it was.